core/modules/replica/Controller.h

```diff
@@ -0,0 +1,687 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_REPLICA_CONTROLLER_H
+#define LSST_QSERV_REPLICA_CONTROLLER_H
+
+/// Controller.h declares:
+///
+/// class ControllerIdentity
+/// class Controller
+///
+/// (see individual class documentation for more information)
+
+// System headers
+#include <map>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+// Third party headers
+#include <boost/asio.hpp>
+
+// Qserv headers
+#include "replica/Request.h"
+#include "replica/RequestTypesFwd.h"
+#include "replica/ServiceProvider.h"
+
+// Forward declarations
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+
+class ControllerImpl;
+class Messenger;
+
+/**
+ * The base class for implementing requests registry as a polymorphic
+ * collection to store active requests. Pure virtual methods of
+ * the class will be overriden by request-type-specific implementations
+ * (see struct RequestWrappeImplr<REQUEST_TYPE> in the .cc file) capturing
+ * type-dependant pointer and a callback function.
+ */
+struct ControllerRequestWrapper {
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<ControllerRequestWrapper> pointer;
+
+    /// Destructor
+    virtual ~ControllerRequestWrapper() = default;
+
+    /// This method will be called upon a completion of a request
+    /// to notify a subscriber on the event.
+    virtual void notify ()=0;
+
+    /// Return a pointer to the stored request object
+    virtual std::shared_ptr<Request> request () const=0;
+};
+
+/**
+ * The data structure encapsulating various attributes which identity
+ * each instance of the Controller class. This information is meant to
+ * be used in the multi-Controller setups to coordinate operations
+ * between multiple instances and to avoid/resolve conflicts.
+ */
+struct ControllerIdentity {
+
+    /// A unique identifier of the Controller
+    std::string id;
+
+    /// The name of a host where it runs
+    std::string host;
+
+    /// An identifier of a process
+    pid_t pid;
+};
+
+/// The overloaded streaming operator for Controller's identity
+std::ostream& operator <<(std::ostream& os, ControllerIdentity const& identity);
+
+/**
+  * Class Controller is used for pushing replication (etc.) requests
+  * to the worker replication services. Only one instance of this class is
+  * allowed per a thread.
+  *
+  * NOTES:
+  *
+  * - all methods launching, stopping or checking status of requests
+  *   require that the server was runinng. Otherwise it will throw
```

Comment:
"require that the server was running" -> "require the server to be running" ?

Reply:
Fixed